--- /usr/lib/scilab-2.7/macros/scicos/Others.cosf	2004-02-05 09:58:02.000000000 +0100
+++ Others.cosf	2004-02-05 09:39:47.000000000 +0100
@@ -2,15 +2,15 @@ scicos_ver = 'scicos2.7'
 scs_m=scicos_diagram()
 scs_m.props=tlist(..
 ['params','wpar','title','tol','tf','context','void1','options','void2','void3',..
-'doc'],[600,450,0,0,600,450],['Others','/usr/local/src/scilab/macros/scicos/'],..
-[0.0001,0.000001,1.000E-10,100001,0,0],100000,' ',[],..
+'doc'],[600,450,0,0,600,450],['Others','/home/bucher/palettes/'],..
+[0.0001,1.000E-06,1.000E-10,100001,0,0],100000,' ',[],..
 tlist(['scsopt','3D','Background','Link','ID','Cmap'],list(%t,33),[8,1],[1,5],..
 list([5,0],[4,0]),[0.8,0.8,0.8]),[],[],list())
 scs_m_1=scicos_diagram()
 scs_m_1.props=tlist(..
 ['params','wpar','title','tol','tf','context','void1','options','void2','void3',..
 'doc'],[600,450,0,0,600,450],'Super Block',..
-[0.0001,0.000001,1.000E-10,100001,0,0],100000,[],[],..
+[0.0001,1.000E-06,1.000E-10,100001,0,0],100000,[],[],..
 tlist(['scsopt','3D','Background','Link','ID','Cmap'],list(%t,33),[8,1],[1,5],..
 list([5,0],[4,0]),[0.8,0.8,0.8]),[],[],list())
 scs_m.objs(1)=mlist(..
@@ -97,7 +97,7 @@ scs_m.objs(7)=mlist(['Block','graphics',
               ['graphics','orig','sz','flip','exprs','pin','pout','pein','peout',..
               'gr_i','id'],[0,70],[40,40],%t,['0';'1'],0,0,0,[],..
               list('xstringb(orig(1),orig(2),''Mem'',sz(1),sz(2),''fill'')',8),..
-              emptystr()),..
+              'Mem'),..
               mlist(..
               ['model','sim','in','out','evtin','evtout','state','dstate','rpar',..
               'ipar','blocktype','firing','dep_ut','label'],'memo',1,1,1,[],[],0,..
